### PR TITLE
add additional columns to source download csv

### LINKED ISSUE #3396: Downloading additional source information
In addition to the current columns provided in the file from a 'Download CSV' request for sources, users may benefit from having the option to download additional information like classification probabilities and annotations. Adding this functionality to downloads via the frontend would complement programmatic access to this information.

### PR DESCRIPTION
**Pull request by jadalilleboe**
Add classification probability, annotation key and annotation value columns to source download csv.
Closes #3396

CSV with demo data:
```
"id","ra [deg]","dec [deg]","redshift","classification","probability","annotation key","annotation value","groups","Date saved","Alias","Origin","TNS Name"
"ZTF18aabcvnq","172.0534465","25.6596908","","Time-domain Source;variable","0.9;0.02","class;some_boolean;numeric_field","varstar;false;0.213471118","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","AT 2018cbj","ZTF",""
"ZTF20abwysqy","6.7856144","34.027306","0.7481","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTF21aaqjmps","202.5050541","13.4111859","0.0246","Galactic Nuclei","0.25","class;some_boolean;numeric_field","SNe II;true;0.11235813","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","SN 2021gvm","ZTF",""
"ZTFJ201825+380316","304.60791628","38.05469","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTFe028h94k","229.9620403","34.8442757","","Galactic Nuclei","0.25","period;period","0.5281475;0.5281217208459","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","SN2020abc","",""
"ZTFinGCN","229.9620403","34.8442757","3","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTFrlh6cyjh","342.4596","51.2634","","","","period","1.107332","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
```

### CODE REVIEW

**Comment by jadalilleboe** [comment]
@bfhealy I request your review, although it seems I can't through github due to repo permissions

**Comment by bfhealy** [comment]
Thanks @jadalilleboe, this will be a useful addition! Would it be possible to add the annotation origins to the download as well? That would help distinguish between annotations where the key is the same (e.g. I've been posting annotations with origins called 'uploadedperiod' and 'revisedperiod', but the key for each annotation is 'period'. The annotation origin is distinct from the source origin, which is already included in the download.

File: static/js/components/SourceTable.jsx [hunk 1]
```diff
@@ -1203,6 +1203,31 @@ const SourceTable = ({
         });
         return classifications.join(";");
       };
+      const renderDownloadProbability = (source) => {
+        const classifications = [];
+        source?.classifications.forEach((x) => {
+          classifications.push(x.probability);
+        });
+        return classifications.join(";");
```
Comment (guynir42) [inline static/js/components/SourceTable.jsx:1211]:
```suggestion
        const probabilities = [];
        source?.classifications.forEach((x) => {
          probabilities.push(x.probability);
        });
        return probabilities.join(";");
```

**Review by guynir42 — APPROVED** [review]
I have a minor suggestion. 
Other than that it looks good. 
Please make sure to test this on real data and look at the results to make sure it does what you think!

**Comment by jadalilleboe** [comment]
@bfhealy like this?
```
"id","ra [deg]","dec [deg]","redshift","classification","probability","annotation origin","annotation key","annotation value","groups","Date saved","Alias","Origin","TNS Name"
"ZTF18aabcvnq","172.0534465","25.6596908","","Time-domain Source;variable","0.9;0.02","Program b","class;some_boolean;numeric_field","varstar;false;0.213471118","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","AT 2018cbj","ZTF",""
"ZTF20abwysqy","6.7856144","34.027306","0.7481","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTF21aaqjmps","202.5050541","13.4111859","0.0246","Galactic Nuclei","0.25","Program A","class;some_boolean;numeric_field","SNe II;true;0.11235813","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","SN 2021gvm","ZTF",""
"ZTFJ201825+380316","304.60791628","38.05469","","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTFe028h94k","229.9620403","34.8442757","","Galactic Nuclei","0.25","Program B;ZVM","period;period","0.5281475;0.5281217208459","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","SN2020abc","",""
"ZTFinGCN","229.9620403","34.8442757","3","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTFrlh6cyjh","342.4596","51.2634","","","","ZVM","period","1.107332","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
```
or should the annotation origin be associated with each key in the annotation keys? I'm thinking it could be confusing if there are multiple key value pairs and you don't know which origin goes with which, but i don't know

**Review by bfhealy — APPROVED** [review]
Thanks @jadalilleboe, this implementation is what I was picturing! I see what you mean about the potential confusion of annotations. I think the current organization of data is appropriate as long as the order of annotation origins/keys/values is preserved during download. There is a similar potential for confusion between classifications and their probabilities, but I think the user will be able to associate the correct fields together.

**Comment by jadalilleboe** [comment]
@bfhealy order is preserved, but what about annotation origins with multiple key value pairs? If annotation1 has 1 key-value pair and annotation2 has 3 key-value pairs, there will be 4 keys in the annotation key column and 4 values in the annotation value column, 2 origins in annotation origin column, but no way to tell which origins map to which key value pairs. With classifications and probabilities it's an ordered 1 to 1 classification-probability mapping right?

**Comment by bfhealy** [comment]
@jadalilleboe That's right - would it be possible to provide duplicate entries in the origin column corresponding to the number of key/value pairs per origin?

**Comment by jadalilleboe** [comment]
```
"id","ra [deg]","dec [deg]","redshift","classification","probability","annotation origin","annotation origin key-value pair count","annotation key","annotation value","groups","Date saved","Alias","Origin","TNS Name"
"ZTF18aabcvnq","172.0534465","25.6596908","","Time-domain Source;variable","0.9;0.02","Program b","3","class;some_boolean;numeric_field","varstar;false;0.213471118","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","AT 2018cbj","ZTF",""
"ZTF20abwysqy","6.7856144","34.027306","0.7481","","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTF21aaqjmps","202.5050541","13.4111859","0.0246","Galactic Nuclei","0.25","Program A","3","class;some_boolean;numeric_field","SNe II;true;0.11235813","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","SN 2021gvm","ZTF",""
"ZTFJ201825+380316","304.60791628","38.05469","","","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTFe028h94k","229.9620403","34.8442757","","Galactic Nuclei","0.25","Program B;ZVM","1;1","period;period","0.5281475;0.5281217208459","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","SN2020abc","",""
"ZTFinGCN","229.9620403","34.8442757","3","","","","","","","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
"ZTFrlh6cyjh","342.4596","51.2634","","","","ZVM","1","period","1.107332","Sitewide Group;Program A;Program B","2022-09-18T14:56:05","","",""
```
added a key-value pair count column for each annotation origin, although the column name is a bit lengthy. thoughts? @bfhealy @guynir42

**Comment by bfhealy** [comment]
Thanks @jadalilleboe, LGTM. I know the column name is a bit long, but it does clear up the ambiguity between origins and keys/values.

**Comment by guynir42** [comment]
@jadalilleboe yeah I think this is ok.

**Pull request merged**